### ISSUE TITLE
[DEV] Use sphere-plane collision tests for movement spells

### ISSUE THREAD

**Issue by Pursche** [azerothcore/forum#24]
This issue is supposed to be a discussion about heightmaps, height checks and navigation.
Movement spells include blink, shadowstep, charge etc.

I have been looking into height testing on WoW terrains and I believe that I have found the reason that spells like Blink sometimes teleports the player through the ground.

The way height testing is implemented serverside right now is that we figure out what terrain triangle the player is standing on (from the players X and Y position), after finding the triangle we get the vertices of the triangle, and then we use interpolation to figure out the height of a certain point inside of that triangle. This can all be seen in GridMap::getHeightFromFloat.

This is a bit of a simplistic implementation of height testing and it does not match up with the client, which I will argue below. If they do not match up and the client thinks we're under the terrain then we will simply fall under the world. To account for this all cores I've seen add an offset to the height to prevent the player from falling through the ground, you can usually see this as an offset of +0.5f or +2.0f depending on the surrounding code. For years now, the standard way of "fixing" issues like Blink falling through the map has been to increase this offset, but this adds the drawback of the player falling a bit after blinking on flat ground.

Now lets look at how the client detects height, please take a look at this image: https://imgur.com/xvNT4Jv

In the chat you can see my equivalence to the .gps command, first it shows the position the client reported and then it shows the height the server calculated on that X and Y coordinate. After that comes some debug stuff, but it's not important for this issue. Please notice how the clientside test and the serverside test doesn't match up and the center of the players feet seem to hover over the ground, this happens because the client uses a more complicated collision test that uses a sphere-plane intersection rather than a triangle interpolation. According to the chat message, if this was the destination of a blink and the height offset (read: hacky fix) for this test is +0.5f we would fall through the ground.

So, to finally fix this I propose keeping the current GetHeight function that uses triangle interpolation, but also adding a GetAccurateHeight function that builds a plane from the triangle the player is standing on and does a sphere-plane intersection test on it. GetAccurateHeight would be used in most places where you want to teleport a player, seeing as a player falling through the ground is a pretty unacceptable bug compared to an NPC slightly clipping into the ground. Keep the current GetHeight function for hot paths like NPC height testing.

Now, the obvious response to all of this is "why don't you fix it?". I would love to fix this issue for AC, but I am not well versed in your code and I have way too many projects on my plate already, and on top of all of this I have quite a lot of extra tasks at work that will require overtime the next 1-2 months.

If you are interested, I would love to sit down with one of your developers, preferably someone who has worked with movement/height testing before, and guide him/her through my reference implementation of a triangle interpolation height test. After that we would work on a sphere-plane collision test (which I have already started work on), until we can get a proof of concept working, at that point it would most likely take him/her a bit of work to seamlessly get it into AC and find all the correct places to use the accurate height test instead of the offset.

Simply put, if you want I can give you a reference implementation and talk you through how it works, but integrating it with ACs existing codebase is outside of the scope of my abilities and the amount of free time I have at the moment.

**Comment by Viste** [comment]
maeby here something usefull: https://github.com/TrinityCore/TrinityCore/commit/9fcbd8f15d249070aabc25d48df86681e8ec3d11

**Comment by Pursche** [comment]
@Viste I'm afraid that's not very useful, from what I can tell TC also only uses the triangle interpolation method, and while that commit mentions GetMapHeight it doesn't really do anything different. It's a more fundamental issue than that.

**Comment by sucofog** [comment]
@Kvipka

**Comment by jackpoz** [comment]
what you are looking for is
- https://github.com/TrinityCore/TrinityCore/commit/6443e518cfa110434617c26f949b6783fe4d10d3
- https://github.com/TrinityCore/TrinityCore/commit/c0676a48f70923994e00629d775f4bfef1007155
- https://github.com/TrinityCore/TrinityCore/commit/16ccc17621313e5c954f26ad23b15392fe06e763
- https://github.com/TrinityCore/TrinityCore/commit/d1080af8db63c78986a7fcc94fcfb2fdfe817734

**Comment by FrancescoBorzi** [comment]
thanks @jackpoz

**Comment by jackpoz** [comment]
it's not perfect (the vmap check still tries to "fix" the position and sometimes teleports few inches backwards) but it's a huge improvement :)

**Comment by Yehonal** [comment]
I'm investigating on it since years. Never noticed this post.
It will be a great improvement.

**Comment by Pursche** [comment]
This might be an improvement but I'm afraid this is in fact **not** what I'm looking for. These are more hack fixes to avoid fixing the underlying issue, not even using detour raycasts will 100% fix the issue.

The server models the creature (or player)-terrain collisions as a point versus a plane, while the client models it as a cylinder versus a plane collision. For 100% replicated functionality, which would _actually_ be proofed against terrain clipping, you can't simplify this issue as a single ray-plane intersection, you'll still need hack fixes or risk falling through the ground.

If you look at the image I posted above, you can clearly see that the center of the character, the place where you would shoot the raycast from, is in the air while the edge of the characters cylinder collider is what is actually colliding with the ground. You just can't get this functionality with ray-plane collisions. If you try to teleport the player to that position he will be lowered to the point where the center of his collider touches the ground, and with a steep enough slope that would allow him to clip through the ground again. This is where all the hack fixes come in, they are meant to instead teleport the player X yards over the intersection point to try to fix the issue, but given a steep enough slope that is still somewhat walkable it would still happen.

**Comment by jackpoz** [comment]
that's not how recastnavigation raycast works though, look at these images:

![image](https://user-images.githubusercontent.com/1153754/75979941-d73d6900-5ee1-11ea-8114-c802f85aa3e3.png)

![image](https://user-images.githubusercontent.com/1153754/75979943-d99fc300-5ee1-11ea-946b-df865092a30d.png)

As you can see End is actually underground (same Z as Start) but the ray is safely above ground. It's easy to then get the end point from the ray, get the Z from mmaps poly and get a more precise one using maps/vmaps.

This in the ingame result at the same position as the recastdemo screenshots above:
![image](https://user-images.githubusercontent.com/1153754/75980094-208db880-5ee2-11ea-9d6a-4169bc81f685.png)

![image](https://user-images.githubusercontent.com/1153754/75980100-22f01280-5ee2-11ea-88b8-5add05289e5d.png)

Give it a try ingame, it might surprise you :)

**Comment by Pursche** [comment]
I know exactly how Recast works, and while this might be an improvement over what you had before, it is still not producing the same results as the client, it's **impossible** for it to do that with just a raycast. Here, let me bring out Paint and explain the issue as simple as I can.

Since my paint skills absolutely suck, let me simplify the issue to just care about one triangle.

![Collision1](https://user-images.githubusercontent.com/4644652/75981196-487e1b80-5ee4-11ea-977a-32c78ddb5a82.png)

Assume that this is the triangle we care about, there is a height value defined for each corner and we have our point P which is where for example we want the player to Blink to.

The client however will not treat P as a point, it will treat P as a cylinder, from a top down perspective it would look something like this.

![Collision2](https://user-images.githubusercontent.com/4644652/75981279-7cf1d780-5ee4-11ea-9cbd-954b33b37900.png)

People seem to assume that this isn't very important and that a simple raycast is "close enough", but this whole mindset is the reason you have layers upon layers of height data adjustments in your codes, all those +0.5f or +2.0f should NOT be needed to prevent terrain clipping, ever.

To visualize this, lets switch over to a side-view perspective.

![Collision3](https://user-images.githubusercontent.com/4644652/75981480-dc4fe780-5ee4-11ea-975d-1f5ba5629822.png)

The red dot represents the position we get from the height test, but the green line represents the players collider. If we teleport the player to the red position he will partially clip through the terrain.

If we asked the client to do the same thing, it would result in a position like this:

![Collision4](https://user-images.githubusercontent.com/4644652/75981555-0a352c00-5ee5-11ea-9b07-8c32ceeab5e3.png)

This example isn't a very big deal though, the client manages to resolve this correctly without any issues. However, imagine that we are standing on a way steeper triangle...

![Collision5](https://user-images.githubusercontent.com/4644652/75981597-22a54680-5ee5-11ea-951b-e38ba56fd65f.png)

In this case, red is the height test we got, blue is the position we teleported the player and green is the position the clients more advanced cylinder test would have picked. In this case, the client would get confused and have a very real chance of falling through the world to its death.

The only way to **actually** solve the underlying issue, rather than hide the effects of it, is to implement a _real_ cylinder-triangle test, or at least a circle-triangle test if we don't care about the players height.

There is absolutely no way a raycast can produce the exact same results as the cylinder-triangle test shown here, you might consider the result you've got "good enough" but I don't think we should settle for good enough, this github issue is about fixing the underlying issue rather than add more ways of hiding it.
